ocaml/tests/test_psr.ml

```diff
@@ -0,0 +1,159 @@
+let checkpoint = ref None
+
+let backups = ref None
+
+let default_pool_secret = "default_pool_secret"
+
+let new_pool_secret = "new_pool_secret"
+
+(* we don't make the distinction between a slave & master here.
+   it is intended that whilst the master orchestrates the procedure
+   in the real implementation, the way the master should tell itself
+   to update the PS is no different to the way the master should tell
+   a slave *)
+module Impl = struct
+  type pool_secret = string
+
+  type pool_secrets = pool_secret * pool_secret
+
+  type fistpoint =
+    | No_fistpoint
+    | Accept_new_pool_secret
+    | Send_new_pool_secret
+    | Finish
+
+  exception Fistpoint of fistpoint
+
+  type host_t = {
+      id: int
+    ; current_pool_secret: string
+    ; staged_pool_secret: string option
+    ; fistpoint: fistpoint
+  }
+
+  type host = host_t ref
+
+  let mk_hosts num =
+    List.init num (fun id ->
+        ref
+          {
+            id
+          ; current_pool_secret= default_pool_secret
+          ; staged_pool_secret= None
+          ; fistpoint= No_fistpoint
+          })
+
+  let save_checkpoint x = checkpoint := Some x
+
+  let retrieve_checkpoint () = !checkpoint
+
+  let backup pool_secrets = backups := Some pool_secrets
+
+  let retrieve () = Option.get !backups
```

Comment:
This will raise an exception if it gets called before `let backup` is called, is it not better to return an option so the caller can deal with it?

Reply:
i wrote `val retrieve : unit -> pool_secrets option` originally, but it actually doesn't make much sense. if we can't fetch the pool secret from the backup that's a fatal error (i.e. the implementor should handle that). the reason it's a fatal error is that we fetch from the backup only when there is a checkpoint to resume from (you'll see that the function to get the checkpoint returns an option)

there are racy, and i ran into this when writing the tests :-1: i think this is somewhat useful here because i do want to make sure that the pool secret backup is cleaned up (for example). i'm not sure what i'd use as an alternative to model writing/reading to a file?